Fixed issue with hiding conditions not hiding containers in simple ticket view

akstis-typer opened the pull request:
- It fixes #847

DaviidMM commented:
Just to  clarify, in my installation (new, all of it with the last versions) I had to place the code in `showForTab` funcion below the line `$display_condition = new PluginFieldsContainerDisplayCondition();`:

```php
...
$display_condition = new PluginFieldsContainerDisplayCondition();

// @akstis-typer code starts here ------
if($item->fields['type'] == "")
{
    $item->fields['type'] = $params['options']['type'];
}
if($item->fields['itilcategories_id'] == "")
{
    $item->fields['itilcategories_id'] = $params['options']['itilcategories_id'];
}
// @akstis-typer code ends here ------

if ($display_condition->computeDisplayContainer($item, $c_id)) {
...
```

DaviidMM commented:
I've been testing this code and works correctly except for one more issue: When a custom field is shown and you edit another field (e.g. title), the custom field's alignment breaks again. My solution was to add `echo "<div class='offset-md-1 col-md-8 col-xxl-6'>";` into the `ajax/container.php` file, inside the `if ($display_condition->computeDisplayContainer($item, $containers_id)) {` condition:

```php
$display_condition = new PluginFieldsContainerDisplayCondition();
    if ($display_condition->computeDisplayContainer($item, $containers_id)) {
        $field_options = [
            'label_class' => 'col-lg-3',
            'input_class' => 'col-lg-9',
        ];
        echo "<div class='offset-md-1 col-md-8 col-xxl-6'>";
        PluginFieldsField::showDomContainer(
            $containers_id,
            $item,
            $type,
            $subtype,
            $field_options
        );
        echo "</div>";
    } else {
        echo '';
    }
```

stonebuzz commented:
Hi @akstis-typer 

can you rebase ?

akstis-typer commented:
> Hi @akstis-typer
> 
> can you rebase ?

Hi, you mean I need to add last changes of original repository? Or something else?
I'm not so good at git so I didn't get it :0

stonebuzz commented:
```
git fetch upstream (github repo)
git pull --rebase upstream main
git push --force 
```

That should be sufficient.

akstis-typer commented:
I think I did it right. The next step is unit tests - how do I make them? Or can I run yours?

trasher commented on composer.json:
```diff
@@ -5,7 +5,6 @@
     },
     "require-dev": {
         "friendsofphp/php-cs-fixer": "^3.75",
-        "friendsoftwig/twigcs": "^6.1",
```
I do not think it should be removed.

Rom1-B replied:
I agree: we have to keep it.

trasher reviewed (commented):
You PR has lot of unrelated changes (new lines added, spacing added/removed) that must be reverted.

akstis-typer commented:
I've done poorly last time and accidentally revert changes in twig files. Now I reverted all back how it should be and removed spaces and new lines I added.

Rom1-B commented on templates/fields.html.twig:
This file has no reason to be modified

Rom1-B commented on templates/status_overrides.html.twig:
This file has no reason to be modified

Rom1-B commented on twig_cs.dist.php:
This file must keep its original name

akstis-typer committed:
Rebased and added only necessary things

Rebased myself to upstream main branch
Edited only two files:
-container.php that fixes formatting of field
-field.class.php that fixes issue with field not respecting hiding condition in simple view

akstis-typer commented:
Finally I got it

leeopereira commented:
=

> I've been testing this code and works correctly except for one more issue: When a custom field is shown and you edit another field (e.g. title), the custom field's alignment breaks again. My solution was to add `echo "<div class='offset-md-1 col-md-8 col-xxl-6'>";` into the `ajax/container.php` file, inside the `if ($display_condition->computeDisplayContainer($item, $containers_id)) {` condition:
> 
> ```
> $display_condition = new PluginFieldsContainerDisplayCondition();
>     if ($display_condition->computeDisplayContainer($item, $containers_id)) {
>         $field_options = [
>             'label_class' => 'col-lg-3',
>             'input_class' => 'col-lg-9',
>         ];
>         echo "<div class='offset-md-1 col-md-8 col-xxl-6'>";
>         PluginFieldsField::showDomContainer(
>             $containers_id,
>             $item,
>             $type,
>             $subtype,
>             $field_options
>         );
>         echo "</div>";
>     } else {
>         echo '';
>     }
> ```
> 
> After the `showDomContainer` method i close the created div with `echo "</div>";`



I tested it and in my environment it only worked correctly this way.

Thank you very much everyone!

leeopereira commented:
Great, congratulations!
You updated the plugin, but when I applied the update, not only did it not fix this problem, but the fix from the colleague above no longer works.
Great!

leeopereira commented:
> Hi, @leeopereira, my fix doesn't work on the nightly version of the plugin?

Oi @akstis-typer , Unfortunately not, I had to restore a backup of the previous version of the files. When I updated, the deleted files were replaced with the originals from the new version, and the problem returned. The hide condition stopped working, and the custom fields became misaligned. :/

akstis-typer commented:
@leeopereira , this happened because this changes are not in the main branch of fields plugin repo, so when you update using glpi - they getting overriden by update that developers made.
But, if you pull changes from this branch and merge it with updated branch - it'll still work :)
I may update my branch to the nightly one, and you'll be able to pull it without problem.

leeopereira commented:
@akstis-typer After updating, I tried editing the files and inserting the corrections here in the topic, but it still didn't work. I don't know if they changed something in the code that impacted the correction. But I just returned the two old edited files and it worked.

Rom1-B commented on ajax/container.php:
```diff
@@ -56,12 +56,19 @@
 
     $display_condition = new PluginFieldsContainerDisplayCondition();
     if ($display_condition->computeDisplayContainer($item, $containers_id)) {
+        $field_options = [
+            'label_class' => 'col-lg-3',
+            'input_class' => 'col-lg-9',
+        ];
+        echo "<div class='offset-md-1 col-md-8 col-xxl-6'>";
         PluginFieldsField::showDomContainer(
             $containers_id,
             $item,
             $type,
             $subtype,
+            $field_options,
         );
+        echo "</div>";
```
This wrapper is now always applied on AJAX refresh, but `showForTab` (`inc/field.class.php:965-974`) only applies it for `helpdesk.public.php`/`tracking.injector.php`. Normal admin/agent forms load with no wrapper, then get re-wrapped as soon as any field changes ; fields shift after the first edit. Matches misalignment reports on the PR thread (e.g. leeopereira, 2025-08-28).

Should this only apply when `$type`/`$subtype` matches the public/injector context, like `showForTab`?

Rom1-B commented on inc/field.class.php:
```diff
@@ -887,6 +887,12 @@ public static function showForTab($params)
     {
         $item = $params['item'];
 
+        if ($item->fields['type'] == "") {
+            $item->fields['type'] = $params['options']['type'];
+        }
+        if ($item->fields['itilcategories_id'] == "") {
+            $item->fields['itilcategories_id'] = $params['options']['itilcategories_id'];
+        }
```
`post_item_form` is a global hook, fired for all 70+ itemtypes, not just `Ticket`. Most have no `type`/`itilcategories_id` column (even `Change`/`Problem` lack `type`), so this raises "Undefined array key" warnings on nearly every item's edit form.

```suggestion
        if ($item instanceof Ticket) {
            $item->fields['type'] = $item->fields['type'] ?: ($params['options']['type'] ?? '');
        }
        if ($item instanceof CommonITILObject) {
            $item->fields['itilcategories_id'] = $item->fields['itilcategories_id'] ?: ($params['options']['itilcategories_id'] ?? '');
        }
```

Can you test with a non-ITIL item (e.g. `Computer`)? No test covers this method.
